error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/Slf4jLogDeclaration.java

```diff
@@ -0,0 +1,154 @@
+import org.jspecify.annotations.Nullable;
+import tech.picnic.errorprone.utils.SourceCode;
+
+/**
+ * A {@link BugChecker} that warns when SLF4J declarations are not canonical.
+ *
+ * @apiNote The default canonical logger name can be overridden through {@link ErrorProneFlags flag
+ *     arguments}.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Make sure SLF4J log declarations follow the SLF4J conventions inside classes.",
+    link = BUG_PATTERNS_BASE_URL + "Slf4jLogDeclaration",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = STYLE)
+// XXX: Do not match on `Class` but on `VariableTree`. That improves on the reporting.
+public final class Slf4jLogDeclaration extends BugChecker implements VariableTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Matcher<Tree> LOGGER = isSubtypeOf("org.slf4j.Logger");
+  private static final Matcher<ExpressionTree> GET_LOGGER_METHOD =
+      staticMethod().onDescendantOf("org.slf4j.LoggerFactory").named("getLogger");
+
+  private static final String CANONICAL_LOGGER_NAME_FLAG =
+      "Slf4jLogDeclaration:CanonicalLoggerName";
+  private static final String DEFAULT_CANONICAL_LOGGER_NAME = "LOG";
+  private static final Pattern STRING_LITERAL_ARGUMENT = Pattern.compile("\"(.*?)\"");
+  private static final Pattern CLASS_ARGUMENT = Pattern.compile("(.*?)\\.class");
+
+  private final String canonicalLoggerName;
+
+  /** Instantiates a default {@link Slf4jLogDeclaration} instance. */
+  public Slf4jLogDeclaration() {
+    this(ErrorProneFlags.empty());
+  }
+
+  /**
+   * Instantiates a customized {@link Slf4jLogDeclaration}.
+   *
+   * @param flags Any provided command line flags.
+   */
+  @Inject
+  Slf4jLogDeclaration(ErrorProneFlags flags) {
+    canonicalLoggerName = getCanonicalLoggerName(flags);
+  }
+
+  @Override
+  public Description matchVariable(VariableTree tree, VisitorState state) {
+    if (!LOGGER.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    Symbol enclosingElement = ASTHelpers.getSymbol(tree).getEnclosingElement();
+    if (enclosingElement == null) {
+      return Description.NO_MATCH;
+    }
+
+    SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+
+    if (enclosingElement.getKind() != ElementKind.INTERFACE) {
+      suggestCanonicalModifiers(tree, fixBuilder, state);
+    }
+    canonicalizeLoggerVariable(tree, fixBuilder, state);
+    updateGetLoggerArgument(tree, enclosingElement.getSimpleName(), fixBuilder, state);
+
+    return fixBuilder.isEmpty() ? Description.NO_MATCH : describeMatch(tree, fixBuilder.build());
+  }
+
+  private void canonicalizeLoggerVariable(
+      VariableTree variableTree, SuggestedFix.Builder fixBuilder, VisitorState state) {
+    if (!variableTree.getName().contentEquals(canonicalLoggerName)) {
+      fixBuilder.merge(SuggestedFixes.renameVariable(variableTree, canonicalLoggerName, state));
+    }
+  }
+
+  private static void updateGetLoggerArgument(
```

Comment:
I'm not seeing why we need a `TreeScanner` to walk over a `VariableTree`, I think we can improve this :).

Reply:
We do this because we need to match on the method invocation of the variable. In case the variable is in a nested class, without this, the match will be on the outer class level (Which was flagged [here](https://github.com/PicnicSupermarket/error-prone-support/blob/14f46b59a8dccbec453b836cbca08a62963fc451/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/Slf4jLogDeclaration.java#L48)).